compiler/rustc_mir_transform/src/check_alignment.rs

```diff
@@ -0,0 +1,274 @@
+use crate::MirPass;
+use rustc_hir::def_id::DefId;
+use rustc_middle::mir::interpret::{ConstValue, Scalar};
+use rustc_middle::mir::visit::{PlaceContext, Visitor};
+use rustc_middle::mir::*;
+use rustc_middle::ty::Ty;
+use rustc_middle::ty::{TyCtxt, TypeAndMut};
+
+pub struct CheckAlignment;
+
+struct PointerFinder<'tcx, 'a> {
+    local_decls: &'a mut LocalDecls<'tcx>,
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    pointers: Vec<(Place<'tcx>, Ty<'tcx>)>,
+}
+
+impl<'tcx, 'a> Visitor<'tcx> for PointerFinder<'tcx, 'a> {
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, _location: Location) {
+        if let PlaceContext::NonUse(_) = context {
+            return;
+        }
+        if !place.is_indirect() {
+            return;
+        }
+
+        let the_pointer = Place::from(place.local);
+        let the_pointer_ty = the_pointer.ty(&*self.local_decls, self.tcx).ty;
+
+        // We only want to check unsafe pointers
+        if !the_pointer_ty.is_unsafe_ptr() {
+            trace!("Indirect, but not an unsafe ptr, not checking {:?}", the_pointer_ty);
+            return;
+        }
+
+        let Some(pointee) = the_pointer_ty.builtin_deref(true) else {
+            debug!("Indirect but no builtin deref: {:?}", the_pointer_ty);
+            return;
+        };
+        let mut pointee_ty = pointee.ty;
+        if pointee_ty.is_array() || pointee_ty.is_slice() || pointee_ty.is_str() {
+            pointee_ty = pointee_ty.sequence_element_type(self.tcx);
+        }
+
+        if !pointee_ty.is_sized(self.tcx, self.tcx.param_env_reveal_all_normalized(self.def_id)) {
+            debug!("Unsafe pointer, but unsized: {:?}", the_pointer_ty);
+            return;
+        }
+
+        self.pointers.push((the_pointer, pointee_ty))
+    }
+}
+
+impl<'tcx, 'a> PointerFinder<'tcx, 'a> {
+    fn find_pointers(&mut self, statement: &Statement<'tcx>) -> Vec<(Place<'tcx>, Ty<'tcx>)> {
+        self.pointers.clear();
+        self.visit_statement(statement, Location::START);
+        core::mem::take(&mut self.pointers)
+    }
+}
+
+impl<'tcx> MirPass<'tcx> for CheckAlignment {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        // Skip this pass if debug-assertions are off
+        if !tcx.sess.opts.debug_assertions {
+            return;
+        }
+
+        let basic_blocks = body.basic_blocks.as_mut();
+        let local_decls = &mut body.local_decls;
+
+        // If the assertion fails we will insert a new terminator. So this check inserts new basic
+        // blocks. So when we iterate over the basic blocks for this Body, we need to use this loop
+        // structure.
```

Comment:
So... by using MirPatch, you would not need `PointerFinder`, as you could register all the mutations you want in a visitor itself and then have them get applied at the end. I think that would simplify a few things. You'd probably not even need to manually iterate over statements and blocks.

Reply:
How do I use MirPatch to insert a new block in the middle of an existing block?

Reply:
Ideally we'd add a helper for this. Basically it would create a new block, replace its terminator with the current block's terminator, copy all statements starting at the current one over, delete them here, replace the current block's terminator.

Reply:
I can't come up with a satisfying way to implement this transformation in `MirPatch`.

For example, how do I deal with the order of operations issue with adding statements: Do I add the statements before I insert the new block, or after? Same deal goes for patching terminators: Do I patch the terminator of a basic block before inserting new blocks into it?

There is also ambiguity that can arise from interleaved new block insertion plus adding new statements. If we want to insert a new block, then add statements, then add a new block in the middle of the added statements, we need to understand the order those method calls happened in order to have the new statements end up in the right block once the patch is fully applied.

It really seems to me like the whole design of `MirPatch` is that things don't become ambiguous if all the operations are purely additive